pallets/parachain-system/src/validate_block/trie_cache.rs

```diff
@@ -0,0 +1,110 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate alloc;
+
+use core::cell::RefCell;
+use hash_db::{HashDB, Hasher};
+use hashbrown::hash_map::{Entry, HashMap};
+use sp_state_machine::{TrieBackendStorage, TrieCacheProvider};
+use sp_std::boxed::Box;
+use sp_trie::NodeCodec;
+use trie_db::{node::NodeOwned, TrieCache, TrieError};
+
+/// Special purpose trie cache implementation that is able to cache an unlimited number
+/// of values. To be used in `validate_block` to serve values and nodes that
+/// have already been loaded and decoded from the storage proof.
+pub(crate) struct SimpleTrieCache<'a, H: Hasher> {
+	node_cache: core::cell::RefMut<'a, HashMap<H::Out, NodeOwned<H::Out>>>,
+	value_cache: core::cell::RefMut<'a, HashMap<Box<[u8]>, trie_db::CachedValue<H::Out>>>,
```

Comment:
```suggestion
	node_cache: RefMut<'a, BTreeMap<H::Out, NodeOwned<H::Out>>>,
	value_cache: RefMut<'a, BTreeMap<Box<[u8]>, trie_db::CachedValue<H::Out>>>,
```

We should use `BTreeMap` to be dos resistant as we have no randomness to initialize the hash map.